src/serialisation.rs

```diff
@@ -0,0 +1,374 @@
+            MessageKind::NodeMessage => {
+                let node_msg: node::NodeMessage =
+                    rmp_serde::from_slice(&self.payload).map_err(|err| {
+                        Error::FailedToParse(format!("Node message payload as Msgpack: {:?}", err))
+                    })?;
+                Ok(MessageType::NodeMessage(node_msg))
+            }
+        }
+    }
+
+    // The following functions are just for convinience, which allow users to
+    // not needing to create an instance of WireMsg beforehand.
+
+    /// Convenience function which creates a temporary WireMsg from the provided
+    /// bytes, returning the deserialised message.
+    pub fn deserialise(bytes: Bytes) -> Result<MessageType> {
+        Self::from(bytes)?.to_message()
+    }
+
+    /// Convenience function which creates a temporary WireMsg from the provided
+    /// MsgEnvelope, returning the serialised WireMsg.
+    pub fn serialise_infrastructure_query(query: &infrastructure::Query) -> Result<Bytes> {
+        Self::new_infrastructure_query(query)?.serialise()
+    }
+
+    /// Convenience function which creates a temporary WireMsg from the provided
+    /// MsgEnvelope, returning the serialised WireMsg.
+    pub fn serialise_client_msg(msg: &client::MsgEnvelope) -> Result<Bytes> {
+        Self::new_client_msg(msg)?.serialise()
+    }
+
+    /// Convenience function which creates a temporary WireMsg from the provided
+    /// node::Messsage, returning the serialised WireMsg.
+    pub fn serialise_node_msg(msg: &node::NodeMessage) -> Result<Bytes> {
+        Self::new_node_msg(msg)?.serialise()
+    }
+
+    // Private function which returns the bytes size of this WireMsg
+    // taking into account current self-contained payload.
+    fn size(&self) -> usize {
+        WireMsgHeader::size() + self.payload.len()
+    }
+}
+
+// Type of message being sent over the wire, this type
+// is part of the WireMsgHeader.
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum MessageKind {
+    Ping,
+    InfrastructureQuery,
+    ClientMessage,
+    NodeMessage,
+}
+
+impl TryFrom<u8> for MessageKind {
+    type Error = super::Error;
+
+    fn try_from(input: u8) -> Result<Self, Self::Error> {
+        match input {
+            0 => Ok(Self::Ping),
+            1 => Ok(Self::InfrastructureQuery),
+            2 => Ok(Self::ClientMessage),
+            3 => Ok(Self::NodeMessage),
+            other => Err(Error::UnsupportedMessageKind(other)),
+        }
+    }
+}
+
+impl From<MessageKind> for u8 {
+    fn from(kind: MessageKind) -> u8 {
+        match kind {
+            MessageKind::Ping => 0,
+            MessageKind::InfrastructureQuery => 1,
+            MessageKind::ClientMessage => 2,
+            MessageKind::NodeMessage => 3,
+        }
+    }
+}
+
+// Header to be serialisied at the front of the wire message.
+// This header contains the information needed to deserialise the payload.
+#[derive(Debug, PartialEq)]
+struct WireMsgHeader {
+    header_size: u16,
```

Comment:
do we really need this field?
you can just make a helper function to return a size constant?

Reply:
yes we need it to allow for backward compatibility, imagine an app built with v1 of this crate reading messages created with v2 of this crate which adds fields to the header, in that case it can find the payload and obviously the payload will also need to be support backward compatibility.

Reply:
ah, right, ok then.